### ISSUE TITLE
[RPC] Should we support users _not_ calling rpc.shutdown()?

### ISSUE THREAD

**Issue by lw** [pytorch/pytorch#41561]
In https://github.com/pytorch/pytorch/issues/41474 we saw that due to global static variables being destroyed in a certain order, not calling `rpc.shutdown()` explicitly before terminating a program could cause a SIGABRT. The problem comes from a event loop thread of TensorPipe trying to acquire a glog mutex that may be already gone. Avoiding this might be complicated, so before we attempt it, let's clarify whether this is even a supported scenario or whether we require users to always shut down explicitly.

**Comment by mrshenli** [comment]
I'd vote for an explicit `shutdown`, at least for now. Without the `shutdown`, I guess we need a strong assumption from the application: "I have done all my work, and I know no other peers would try to contact me", which would be hard to debug and doesn't seem to offer much benefit?

**Comment by pritamdamania87** [comment]
It looks like TensorPipeAgent calls `shutdown` as part of its destructor: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L239. If so, shouldn't `shutdown` already be called before any global static variables are destroyed?

In general to avoid the static initialization order fiasco, we do use leaky singletons in other places. Ex: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/context/container.cpp#L92

**Comment by Flamefire** [comment]
I'm the one who discovered the issue, let me try to clarify:

> If so, shouldn't shutdown already be called before any global static variables are destroyed?

I'm assuming TensorPipeAgent is the instance listening for requests or at least starting threads doing that. Then the destructor of that is to late. I mean: Isn't TensorPipeAgent and it's threads exactly what should be terminated by `rpc.shutdown()`?   
Basically there are 2 destructions racing: The one from a mutex in glog and the TensorPipeAgent which seems to also be a global static, wrapped in a shared_ptr as mentioned by @lw in https://github.com/pytorch/pytorch/issues/41474#issuecomment-659348539.   
As the problematic global static is in glog there is no control over it (e.g. to make it leaky) unless you get glog to change it.

Other solutions are to force the destruction order by carefully orchestrating the construction order, see https://github.com/pytorch/pytorch/issues/41474#issuecomment-659906084, or to avoid the global static in the first place and require to hold an instance instead. As the instance has to be created in main, it will be destroyed before any global static.

**Comment by lw** [comment]
> It looks like TensorPipeAgent calls shutdown as part of its destructor: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/rpc/tensorpipe_agent.cpp#L239. If so, shouldn't shutdown already be called before any global static variables are destroyed?

Maybe we're mixing two things. There is the `shutdown()` function on the `rpc` module, and then there is the `shutdown` method on the RPC agents. The `rpc.shutdown()` function does call the `RPCAgent::shutdown()` but it does other things too, like reset the global static `shared_ptr` for the current RPC agent.

If `rpc.shutdown()` is never called, the `shared_ptr` is never reset and `RPCAgent::shutdown()` is never called. This means that the destructor of the agent will be called when the global static `shared_ptr` is destroyed. At that point, yes, the agent will call its `RPCAgent::shutdown()` but some other global static objects (like the glog mutexes) may already be gone.

> In general to avoid the static initialization order fiasco, we do use leaky singletons in other places.

Yes, that would be one approach, but as I'm not very familiar with those intricacies of C++ I'm not sure how a solution could look like. One idea, not sure if it works, is to make the `shared_ptr` holding the current RPC agent a _local_ static object of some function. This way, it would only be created when first calling that function. And, if that function, before creating the `shared_ptr`, were to do something that initializes the glog mutexes, then perhaps this would be enough to ensure the correct inverse destruction order. Does that make sense?

The downside is that the global will then be local to a single function, and thus we must merge the `setCurrentRpcAgent`, `getCurrentRpcAgent` and `isCurrentRpcAgentSet` functions (or reimplement them on top of a single function). The other problem is that I'm not sure what we can do to initialize glog except for logging a line, which would the pollute the logs of all users at every run. There may be advanced features of glog for that but I don't know if we can use them since PyTorch should support being built without glog.

**Comment by lw** [comment]
Actually, since the glog mutex is a _global_ static, maybe just have the RPC agent become a _local_ static would be enough, even without its containing function doing anything with glog?

**Comment by Flamefire** [comment]
> One idea, not sure if it works, is to make the shared_ptr holding the current RPC agent a local static object of some function. This way, it would only be created when first calling that function. And, if that function, before creating the shared_ptr, were to do something that initializes the glog mutexes, then perhaps this would be enough to ensure the correct inverse destruction order. Does that make sense?

Yes. Note that the downside isn't that big as you say. The described usage is a very basic example of singleton usage with the benefit of delayed initialization:
```
T& getT(){
  T inst;
  return inst;
}
```

Note that C++ requires the access to be protected by a hidden mutex & lock to ensure it is initialized exactly once. But as most implementations do double-checked locking it should not be noticeable.

I found an issue in glog: https://github.com/google/glog/issues/547 But no responses so far.

> The other problem is that I'm not sure what we can do to initialize glog except for logging a line, which would the pollute the logs of all users at every run

Checking the source where that lock is acquired you can also call many other functions. E.g. `GetLogger` would suffice and not cause any side effects.

**Comment by lw** [comment]
@Flamefire I've attempted that in #41591. Could you revert the previous fix we had for those flaky tests, check out that PR and see if it works or if the flakiness persists? Thanks!

**Comment by jjlilley** [comment]
> is to make the shared_ptr holding the current RPC agent a local static object of some function. This way, it would only be
> created when first calling that function

The `shared_ptr<>` is only lazily set non-null when first calling the function, but isn't the scope/lifetime of the var equivalent to any other `static` module variable? (meaning it's basically like any other static in the module)

It seems like the core issue is the cross-module destructor dependency, which is usually solved by intentionally leaking or removing the dependency.

**Comment by Flamefire** [comment]
No, the `shared_ptr` is actually constructed on first calling the function and hence scheduled for destruction only then. In other words the lifetime of that `shared_ptr` doesn't even begin unless the function is called. See e.g. https://stackoverflow.com/questions/246564/what-is-the-lifetime-of-a-static-variable-in-a-c-function

So I guess the approach in #41591 can work. Rebuilding ATM to verify.

**Comment by jjlilley** [comment]
Interesting! I didn't realize that the c++ runtime actually "remembers" the singleton construction order within a module.  It's possible to verify this with a short program:

```
// clang++ thisfile.cpp
class X {
public:
    explicit X(int v) : v_(v) {}
    ~X() { std::cout << "Destroying X " << v_ << std::endl; }
private:
    int v_;
};

static X* singleton1() {  static X x(1); return &x; }
static X* singleton2() {  static X x(2); return &x; }
static X* singleton3() {  static X x(3); return &x; }

int main(int argc, char** argv) {
    X* x1 = singleton1();    //  *** Reordering these three lines changes the destruction order on module destruction.
    X* x2 = singleton2();
    X* x3 = singleton3();
    (void)x1;(void)x2;(void)x3;
}
```

**Comment by Flamefire** [comment]
#41591 failed :( Same issue

2 reasons I could imagine:
- some other global static graps a reference to the singleton (even indirectly through the other functions)
- cross-module destruction order is the problem. As glog is linked dynamically this might be the issue: The library gets loaded later.

Or. Technically C++ doesn't require to construct a global static until it is used. It usually does but it is not a requirement. Forgot the details/reasoning but might be with shared libraries...

**Comment by lw** [comment]
Shame it didn't work...

The hunch about dynamic linking might be right, and it could explain why I wasn't seeing this issue in my build, as there I link glog statically. I'm a bit out of my depth here, but if you want we can give this one last try. I've updated #41591 so that before initializing the static shared_ptr we do the `GetLogger` call you mentioned. That might force glog to get loaded and initialized and may introduce the dependency we want also in case of dynamic linking. Or maybe it won't. No idea until we try...

**Comment by lw** [comment]
As a more meta comment, it may be possible that in the end we find a way to fix the technical problem we were encountering, but that solution is becoming more and more complicated as we progress. It would definitely be simpler if we said that not calling `rpc.shutdown()` is undefined behavior and thus allowed the possibility of crashing in that case. @mrshenli seemed to agree that we could do so, what do others think?
